cranelift/wasm/src/code_translator.rs

```diff
@@ -2019,6 +2020,87 @@ pub fn translate_operator<FE: FuncEnvironment + ?Sized>(
         | Operator::F64x2RelaxedMax => {
             return Err(wasm_unsupported!("proposed relaxed-simd operator {:?}", op));
         }
+
+        // TODO(dhil) fixme: merge into the above list.
+        // Function references instructions
+        Operator::ReturnCallRef => {
+            return Err(wasm_unsupported!(
+                "proposed tail-call operator for function references {:?}",
+                op
+            ));
+        }
+        Operator::BrOnNull { relative_depth } => {
+            let r = state.pop1();
+            let (br_destination, inputs) = translate_br_if_args(*relative_depth, state);
+            let is_null = environ.translate_ref_is_null(builder.cursor(), r)?;
+            canonicalise_then_brnz(builder, is_null, br_destination, inputs);
+
+            let next_block = builder.create_block();
+            canonicalise_then_jump(builder, next_block, &[]);
+            builder.seal_block(next_block); // The only predecessor is the current block.
+            builder.switch_to_block(next_block);
+            state.push1(r);
+        }
+        Operator::BrOnNonNull { relative_depth } => {
+            // We write this a bit differently from the spec to avoid an extra
+            // block/branch and the typed accounting thereof. Instead of the
+            // spec's approach, it's described as such:
+            // Peek the value val from the stack.
+            // If val is ref.null ht, then: pop the value val from the stack.
+            // Else: Execute the instruction (br relative_depth).
+            let is_null = environ.translate_ref_is_null(builder.cursor(), state.peek1())?;
+            let (br_destination, inputs) = translate_br_if_args(*relative_depth, state);
+            canonicalise_then_brz(builder, is_null, br_destination, inputs);
+            // In the null case, pop the ref
+            state.pop1();
+            // It seems that we're required to create an unconditional jump for
+            // the non-br case, based on the example of BrIf, but i'm not sure why
```

Comment:
This is due to some vestigial bits from when we had extended basic blocks instead of plain basic blocks. Because we have basic blocks, we can't add new non-block-terminating instructions after a `brz`. But also we don't currently have a two-way conditional branch instruction, so blocks ending in control flow splits have to end with something like `brz L1; jump L2` instead. We have general plans to clean this stuff up in the not super distant future, but thats the state of affairs for now.